test/functional/placeholder.js

```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright 2015 The Incremental DOM Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+var IncrementalDOM = require('../../index'),
+    patch = IncrementalDOM.patch,
+    elementOpen = IncrementalDOM.elementOpen,
+    elementClose = IncrementalDOM.elementClose,
+    elementVoid = IncrementalDOM.elementVoid,
+    elementPlaceholder = IncrementalDOM.elementPlaceholder;
+
+describe('placeholders', () => {
+  var container;
+
+  beforeEach(() => {
+    container = document.createElement('div');
+    document.body.appendChild(container);
+  });
+
+  afterEach(() => {
+    document.body.removeChild(container);
+  });
+
+  describe('without an existing DOM element', () => {
+    var el;
+
+    beforeEach(() => {
+      patch(container, () => {
+        el = elementPlaceholder('div', 'placeholder', ['class', 'someClass'], 'data-foo', 'bar');
+      });
+    });
+
+    it('should render with the specified tag', () => {
+      expect(el.tagName).to.equal('DIV');
+    });
+
+    it('should render with static attributes', () => {
+      expect(el.className).to.equal('someClass');
+    });
+
+    it('should not render with dynamic attributes', () => {
+      expect(el.getAttribute('data-foo')).not.to.equal('bar');
+    });
+  });
+
+  describe('with an existing DOM element', () => {
+    var div;
+    var el;
+    var child;
+
+    beforeEach(() => {
+      patch(container, function() {
+        div = elementPlaceholder('div', 'placeholder');
+        patch(div, function() {
+          child = elementVoid('div', null, ['id', 'child']);
+        });
+      });
+
+      patch(container, () => {
+        el = elementPlaceholder('span', 'placeholder', ['class', 'someClass'], 'data-foo', 'bar');
+      });
+    });
+
+    it('should keep the element', () => {
+      expect(el).to.equal(div);
+      expect(div.tagName).to.equal('DIV');
+    });
+
+    it('should not set static attributes', () => {
+      expect(div.getAttribute('class')).not.to.equal('someClass');
+    });
+
+    it('should not set dynamic attributes', () => {
+      expect(div.getAttribute('data-foo')).not.to.equal('bar');
```

Comment:
> The pattern of the outside controlling the container while the render function controls everything inside it doesn't quite apply anymore with the recommendation to pull out an updateAttributes function from general DOM alignment.

The point was to give control back to the library using Incremental DOM control, if it wants it. For example, you might not want to do any attribute value caching and take the performance hit in exchange for a reduced memory footprint.

After thinking about it, alignment and attribute diffing are really separate things so they probably shouldn't even be tied together.

> Now we can treat the container as just a root element, and the view/component/instance attached to that root element is able to exist in isolation from a parent. That principle is one of the things that makes Backbone so damn nice: view's don't require a parent to exist and are fully autonomous and useful without one.

In my opinion, the insertion point should not act as the root of a view. That would mean someone external to your view could decide that your root element should be an `h3` instead of a `div`.  And while I think doing that is not the right pattern, Incremental DOM will not enforce it.

> You'd still be able to mutate a DOM node outside of the normal flow with updateAttributes, even inside the same render.

I don't mind Incremental DOM giving people footguns, as it isn't intended to be used directly like that. The DOM API itself gives you plenty and we're just providing a thin wrapper on top of that.

> But I think breaking encapsulation by supporting this as a first class API is a much more harmful. Imagine some subview needs a special class for the correct CSS styles. But now, a parent comes along and tells the subview to remove it's class.

This comes from the fact that you are trying to use this insertion point that the parent should have control over in the subview. If you really want to retain control, the call to `elementPlaceholder` should be moved outside of the control of the parent. For example, instead of:

```javascript
elementOpen('div');
  elementPlaceholder('div', 'subview', ...);
elementClose('div);
```

you could do something like:

```javascript
elementOpen('div');
  subview({
    viewName: 'subview',
    ...
  });
elementClose('div');

// somewhere else
function subview(params) {
  elementPlaceholder('div', params.viewName, ...);
}
```